src/core/dense_set.cc

```diff
@@ -0,0 +1,558 @@
+}
+
+namespace dfly {
+using namespace std;
+
+constexpr size_t kMinSizeShift = 2;
+constexpr size_t kMinSize = 1 << kMinSizeShift;
+constexpr bool kAllowDisplacements = true;
+constexpr bool kFastScan = true;
+
+DenseSet::ChainIterator& DenseSet::ChainIterator::operator++() {
+  if (curr_) {
+    curr_ = curr_->Next();
+  }
+
+  return *this;
+}
+
+DenseSet::ConstChainIterator& DenseSet::ConstChainIterator::operator++() {
+  if (curr_) {
+    curr_ = curr_->Next();
+  }
+
+  return *this;
+}
+
+bool DenseSet::IsEmpty(DenseSet::ChainVectorIterator it) const {
+  return it->IsEmpty();
+}
+
+size_t DenseSet::PushFront(DenseSet::ChainVectorIterator it, void* data) {
+  // if this is an empty list assign the value to the empty placeholder pointer
+  if (it->IsEmpty()) {
+    it->SetObject(data);
+    return ObjectAllocSize(data);
+  }
+
+  // otherwise make a new link and connect it to the front of the list
+  LinkAllocator la(mr());
+  DenseLinkKey* lk = la.allocate(1);
+  la.construct(lk);
+
+  lk->next = *it;
+  it->SetLink(lk);
+  lk->SetObject(data);
+
+  return ObjectAllocSize(data);
+}
+
+void DenseSet::PushFront(DenseSet::ChainVectorIterator it, DenseSet::DensePtr* ptr) {
+  if (it->IsEmpty()) {
+    it->SetObject(ptr->GetObject());
+    if (ptr->IsLink()) {
+      // deallocate the link if it is no longer a link as it is now in an empty list
+      mr()->deallocate(ptr->Raw(), sizeof(DenseLinkKey), alignof(DenseLinkKey));
+    }
+  } else if (ptr->IsLink()) {
+    // if the pointer is already a link then no allocation needed
+    *ptr->Next() = *it;
+    *it = *ptr;
+  } else {
+    // allocate a new link if needed and copy the pointer to the new link
+    LinkAllocator la(mr());
+    DenseLinkKey* lk = la.allocate(1);
+    la.construct(lk);
+
+    lk->next = *it;
+    lk->SetObject(ptr->GetObject());
+    it->SetLink(lk);
+  }
+}
+
+void* DenseSet::PopFront(DenseSet::ChainVectorIterator it) {
+  if (IsEmpty(it)) {
+    return nullptr;
+  }
+
+  void* ret = it->GetObject();
+
+  // if this is the last object in this chain just reset the
+  // DensePtr instead of deleting it
+  if (it->IsObject()) {
+    it->Reset();
+  } else {
+    DCHECK(it->IsLink());
+    // set the current item in the list to the next one & deallocate the link data structure
+    DenseLinkKey* lk = (DenseLinkKey*)it->Raw();
+    *it = lk->next;
+    mr()->deallocate(lk, sizeof(DenseLinkKey), alignof(DenseLinkKey));
+  }
+
+  return ret;
+}
+
+DenseSet::DensePtr* DenseSet::Find(void* needle, DenseSet::ChainVectorIterator list,
+                                   DenseSet::DensePtr** prev_out) {
+  DensePtr* prev = nullptr;
+  ChainIterator it = ChainBegin(list);
+  for (; it != ChainEnd(list); ++it) {
+    if (Equal(*it, needle)) {
+      break;
+    }
+
+    prev = *it;
+  }
+
+  if (!it->IsEmpty() && prev_out != nullptr) {
+    *prev_out = prev;
+  }
+
+  return *it;
+}
+
+DenseSet::ChainIterator DenseSet::Unlink(DenseSet::ChainIterator node,
```

Comment:
Frankly, after looking at the code, I think that ChainIterator abstraction did not contribute much to the clarity of the code that uses it.  We still need to leak the underlying `DensePtr*`  and check all these conditions of the state of inner `DensePtr*`. Did you try to see where it helps you?

Reply:
Regarding the interface of this function - returning both as a result and via pointer is a bit awkward.

Reply:
I agree returning using both forms is awkward originally I designed the interface to be similar to vector::erase but it ended up being more awkward than anticipated since the contained data needs to be returned to be freed